src/IO/GeoJSONReader.php

```diff
@@ -0,0 +1,357 @@
+            throw GeometryIOException::invalidGeoJSON('Missing "Feature.Geometry" attribute.');
+        }
+
+        return $this->readGeometry($feature['GEOMETRY']);
+    }
+
+    /**
+     * @param array $geometry
+     *
+     * @return Geometry
+     *
+     * @throws GeometryIOException
+     * @throws \Brick\Geo\Exception\InvalidGeometryException
+     * @throws \Brick\Geo\Exception\CoordinateSystemException
+     * @throws \Brick\Geo\Exception\UnexpectedGeometryException
+     */
+    protected function readGeometry(array $geometry) : Geometry
+    {
+        // Verify Geometry TYPE
+        if (! array_key_exists('TYPE', $geometry) || ! is_string($geometry['TYPE'])) {
+            throw GeometryIOException::invalidGeoJSON('Missing "Geometry.Type" attribute.');
+        }
+
+        $geoType = $geometry['TYPE'];
+
+        // Verify Geometry COORDINATES
+        if (! array_key_exists('COORDINATES', $geometry) || ! array($geometry['COORDINATES'])) {
+            throw GeometryIOException::invalidGeoJSON('Missing "Geometry.Coordinates" attribute.');
+        }
+
+        $geoCoords = $geometry['COORDINATES'];
+
+        $hasZ = $this->hasZ($geoCoords);
+        $hasM = false;
+        $srid = 0;
+        $isEmpty = empty($geoCoords);
+
+        $cs = new CoordinateSystem($hasZ, $hasM, $srid);
+
+        switch ($geoType) {
+            case 'POINT':
+                if ($isEmpty) {
+                    return new Point($cs);
+                }
+
+                return $this->genPoint($cs, ...$geoCoords);
+
+            case 'MULTIPOINT':
+                if ($isEmpty) {
+                    return new MultiPoint($cs);
+                }
+
+                return $this->genMultiPoint($cs, ...$geoCoords);
+
+            case 'LINESTRING':
+                if ($isEmpty) {
+                    return new LineString($cs);
+                }
+
+                return $this->genLineString($cs, ...$geoCoords);
+
+            case 'MULTILINESTRING':
+                if ($isEmpty) {
+                    return new MultiLineString($cs);
+                }
+
+                return $this->genMultiLineString($cs, ...$geoCoords);
+
+            case 'POLYGON':
+                if ($isEmpty) {
+                    return new Polygon($cs);
+                }
+
+                return $this->genPolygon($cs, ...$geoCoords);
+
+            case 'MULTIPOLYGON':
+                if ($isEmpty) {
+                    return new MultiPolygon($cs);
+                }
+
+                return $this->genMultiPolygon($cs, ...$geoCoords);
+        }
+
+        throw GeometryIOException::unsupportedGeoJSONType($geoType);
+    }
+
+    /**
+     * [x, y]
+     *s
+     * @param CoordinateSystem $cs
+     * @param float[]          ...$coords
+     *
+     * @return Point
+     *
+     * @throws \Brick\Geo\Exception\InvalidGeometryException
+     */
+    private function genPoint(CoordinateSystem $cs, float ...$coords) : Point
```

Comment:
Remaining Read{Geometry} Functions Type Hinted. [commit: 5c024b2]

Please let me know if I am missing anything here.  I am not sure what you mean by the following code example... OR maybe I do not understand what you mean when it comes to: 
```
it does not make much sense pack/unpack these coordinates arrays throughout the code:

function genLineString(CoordinateSystem $cs, ...$coords)
genPoint($cs, ...$coords)
```

Reply:
My code sample was showing what was currently happening, not what I wanted :)

I meant that you should replace method signatures with:

```php
function genLineString(CoordinateSystem $cs, array $coords)
```

and call them with the $coords array directly:

```php
genPoint($cs, $coords);
```

instead of the above.

Reply:
Resolved in commit: 86180cf